magma/lib/magma/loader.rb

```diff
@@ -134,6 +136,92 @@ def explicit_revision_exists?(revisions:, parent_model:, parent_record_name:, ch
+      # empty list is returned.
+      #
+      # This method succeeds at finding parent identifiers if :
+      # 1. A proper grammar is created with gnomon and the grammar rules correspond to a model hierarchy.
+      # 2. All models have been created in the db.
+      #
+      # The identifiers DO NOT need to exist in the identifier table.
+
+      return {} if @grammar.nil?
+      grammar_decomposed = @grammar.decompose(identifier)
+      return {} if grammar_decomposed.nil?
+
+      # First we gather all available parent models and identifiers associated with the identifier.
+      # If someone has created a well formatted grammar, with proper identifiers,
+      # there MAY be a model hierarchy chain we can infer
+
+      project = Magma.instance.get_project(@project_name)
+      available_models = project.models.keys & grammar_decomposed[:rules].keys.map(&:to_sym)
+
+      chain = {}
+      available_models.each do |model_name|
+        model = project.models[model_name]
+        chain[model_name] = {
+            model: model,
+            identifier_name: model.identity.name,
+            identifier: grammar_decomposed[:rules][model_name.to_s][:name],
+            parent_model_name: model.parent_model_name
+          }
+      end
+
+      # Then we check to make sure a valid hierarchy chain exists
+      # We check that each parent_model is in the list of models available
+      valid = chain.all? do |model_name, model|
+        model_name == :project || chain.has_key?(model[:parent_model_name])
+      end
+
+      valid ? chain : {}
+
+    end
+
+    def autolink_parent_identifiers
+      # If a project is configured with the 'identifier' flag, anytime a record with an identifier is created or updated,
+      # we can implicitly infer what their parents must be via find_parent_models(). We then update the record and its revision
+      # to include its foreign key, and do this up the model hierarchy.
+      #
+      # Note that find_parent_models() is called on each record. If there are two or more records, each with an identifier that
+      # is able to decompose into a model hierarchy - we add each hierarchy to a set. We are ultimately left with a set
+      # that contains the deepest hierarchy.
+      #
+      # It is also important to note that this method is OPTIMISTIC. That is, even though we've inferred a model
+      # hierarchy from a grammar, the identifiers may not exist in the db, and we will only know once we call validate().
+      # Ideally we would validate() before we do this work, however the control flow of this classes makes it hard to do so.
+      # TODO: eventually refactor to address the above comment.
+      #
+      gnomon_mode = Magma::Flags::GNOMON_MODE
+      flag_value = @flags[gnomon_mode[:name]]
+      updates = Set.new([])
+
+      @records.each do |model, record_set|
+        record_set.each do |record_name, record|
+
+          next unless flag_value == gnomon_mode[:identifier]
```

Comment:
@graft note that here, I've dropped the conditional here - previously `record.includes_parent_record?`.

I think this is more reasonable, that we just and build any model hierarchy we can - regardless if it is with the grand-parent, parent or child record. The `updates` var avoids duplicates